services/rsources/handler.go

```diff
@@ -103,78 +107,130 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
 	if sh.config.SkipFailedRecordsCollection {
-		return
+		return nil
 	}
-	stmt, err := tx.Prepare(`insert into "rsources_failed_keys" (
-		job_run_id,
-		task_run_id,
-		source_id,
-		destination_id,
-		record_id
-	) values ($1, $2, $3, $4, $5)`)
+
+	row := tx.QueryRow(`WITH new_key AS (
+		INSERT INTO rsources_failed_keys_v2 (id, job_run_id, task_run_id, source_id, destination_id)
+		VALUES ($1, $2, $3, $4, $5) 
+		ON CONFLICT (job_run_id, task_run_id, source_id, destination_id, db_name) DO NOTHING
+		RETURNING id
+	) SELECT COALESCE(
+		(SELECT id FROM new_key),
+		(SELECT id FROM rsources_failed_keys_v2 WHERE job_run_id = $2 AND task_run_id = $3 AND source_id = $4 AND destination_id = $5)
+	)`, ksuid.New().String(), jobRunId, key.TaskRunID, key.SourceID, key.DestinationID)
+	var id string
+	if err := row.Scan(&id); err != nil {
+		return fmt.Errorf("scanning rsources_failed_keys_v2 id: %w", err)
+	}
+
+	stmt, err := tx.Prepare(`INSERT INTO rsources_failed_keys_v2_records (id, record_id) VALUES ($1, $2) ON CONFLICT DO NOTHING`)
 	if err != nil {
 		return err
 	}
 	defer func() { _ = stmt.Close() }()
 
 	for i := range records {
-		_, err = stmt.ExecContext(
+		if _, err = stmt.ExecContext(
 			ctx,
-			jobRunId,
-			key.TaskRunID,
-			key.SourceID,
-			key.DestinationID,
-			records[i])
-		if err != nil {
-			return err
+			id,
+			records[i]); err != nil {
+			return fmt.Errorf("inserting into rsources_failed_keys_v2_records: %w", err)
 		}
 	}
-
-	return
+	return nil
 }
 
-func (sh *sourcesHandler) GetFailedRecords(ctx context.Context, jobRunId string, filter JobFilter) (JobFailedRecords, error) {
+func (sh *sourcesHandler) GetFailedRecords(ctx context.Context, jobRunId string, filter JobFilter, paging PagingInfo) (JobFailedRecords, error) {
 	if sh.config.SkipFailedRecordsCollection {
 		return JobFailedRecords{ID: jobRunId}, ErrOperationNotSupported
 	}
-	filters, filterParams := sqlFilters(jobRunId, filter)
+	// first find the list of ids (postgres query planner uses an inefficient plan if there is one id with millions of records and a few ids with a few records)
+	ids, err := func() ([]string, error) {
+		var ids []string
+		filters, params := sqlFilters(jobRunId, filter)
+		rows, err := sh.readDB().QueryContext(ctx, `SELECT id FROM "rsources_failed_keys_v2" `+filters, params...)
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = rows.Close() }()
+		for rows.Next() {
+			var id string
+			if err := rows.Scan(&id); err != nil {
+				return nil, err
+			}
+			ids = append(ids, id)
+		}
+		if err := rows.Err(); err != nil {
+			return nil, err
+		}
+		return ids, nil
+	}()
+	if err != nil {
+		return JobFailedRecords{ID: jobRunId}, fmt.Errorf("failed to get failed record ids: %w", err)
+	}
 
+	filters := "WHERE r.id = ANY($1)"
+	params := []interface{}{pq.Array(ids)}
+	var limit string
+	if paging.Size > 0 {
+		if nextPageToken, err := NextPageTokenFromString(paging.NextPageToken); err == nil {
```

Comment:
Minute:
This unmarshal could've been done in the httphandler and return early on failure(and avoid above `SELECT id FROM "rsources_failed_keys_v2"` query).